crates/bevy_ecs/src/schedule/ambiguity_detection.rs

```diff
@@ -0,0 +1,640 @@
+use crate::component::ComponentId;
+use crate::schedule::{AmbiguityDetection, SystemContainer, SystemStage};
+use crate::world::World;
+
+use fixedbitset::FixedBitSet;
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+/// Systems that access the same Component or Resource within the same stage
+/// risk an ambiguous order that could result in logic bugs, unless they have an
+/// explicit execution ordering constraint between them.
+///
+/// This occurs because, in the absence of explicit constraints, systems are executed in
+/// an unstable, arbitrary order within each stage that may vary between runs and frames.
+///
+/// Some ambiguities reported by the ambiguity checker may be warranted (to allow two systems to run
+/// without blocking each other) or spurious, as the exact combination of archetypes used may
+/// prevent them from ever conflicting during actual gameplay. You can resolve the warnings produced
+/// by the ambiguity checker by adding `.before` or `.after` to one of the conflicting systems
+/// referencing the other system to force a specific ordering.
+///
+/// The checker may report a system more times than the amount of constraints it would actually need
+/// to have unambiguous order with regards to a group of already-constrained systems.
+///
+/// By default, the value of this resource is set to `Minimal`.
+///
+/// ## Example
+/// ```ignore
+/// # use bevy_app::App;
+/// # use bevy_ecs::schedule::ReportExecutionOrderAmbiguities;
+/// App::new()
+///    .insert_resource(ReportExecutionOrderAmbiguities::verbose().ignore(&["my_external_crate"]));
+/// ```
+pub enum ReportExecutionOrderAmbiguities {
+    /// Disables all messages reported by the ambiguity checker
+    Off,
+    /// Displays only the number of unresolved ambiguities detected by the ambiguity checker
+    Minimal,
+    /// Displays a full report of ambiguities detected by the ambiguity checker
+    Verbose,
+    /// Verbosely reports all non-ignored ambiguities, including those between Bevy's systems
+    ///
+    /// These will not be actionable: you should only turn on this functionality when
+    /// investigating to see if there's a Bevy bug or working on the engine itself.
+    ReportInternal,
+    /// Verbosely reports ALL ambiguities, even ignored ones
+    ///
+    /// This will be very noisy, but can be useful when attempting to track down subtle determinism issues,
+    /// as you might need when attempting to implement lockstep networking.
+    Deterministic,
+}
+
+/// A pair of systems that can run in an ambiguous order
+///
+/// Created by applying [`find_ambiguities`] to a [`SystemContainer`].
+/// These can be reported by configuring the [`ReportExecutionOrderAmbiguities`] resource.
+#[derive(Debug, Clone, PartialEq)]
+struct SystemOrderAmbiguity {
```

Comment:
While this could in theory by useful to the end user, they can't construct ComponentIds on their own, and the indexes are non-deterministic between runs.

Ditto for `find_ambiguities` and `SystemStage::ambiguities`. See Future Work section of PR description for more.